### PR TITLE
[MODDICONV-286] Allow Job Profiles To Be Imported

### PR DESCRIPTION
**Pull request by RomanChernetskyi**
## Purpose
Allow job profiles to be imported into a FOLIO cluster.
## Approach
- Created dedicated endpoint exists for importing JobProfile - POST data-import-profiles/profileSnapshots
- mod-di-cs will accept input that have a json schema from the output of data-import-profiles/profileSnapshots/{id}?profileType=JOB_PROFILE&jobProfileId={id}
- Job profile and all the linked profiles are created along with appropriate associations between wrappers over those profiles. Exporting that profile should result in exactly the same profileSnapshot structure that was received as an input for importing.
## Learning
Jira: https://folio-org.atlassian.net/browse/MODDICONV-286
Overlay of existing profiles will be implemented in scope of: https://folio-org.atlassian.net/browse/MODDICONV-392

### CODE REVIEW

**Review by dmytrokrutii — APPROVED** [review]

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    try {
+      convertProfileSnapshotWrapperContent(profileSnapshot);
+    } catch (Exception e) {
+      return Future.failedFuture(new BadRequestException(String.format("Cannot map profile content, error: %s", e.getMessage())));
+    }
+
+    var profileTypeToSnapshots = getProfileTypeToSnapshot(profileSnapshot);
+    var profileTypesInSaveOrder = profileTypeToSnapshots.keySet().stream()
+      .sorted(Comparator.comparingInt(IMPORT_PROFILES_ORDER::indexOf))
+      .toList();
```
Comment (VRohach) [inline mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java:92]: @RomanChernetskyi Let's create a separate method for it. For example "orderProfilesBeforeSaving"

File: mod-di-converter-storage-server/src/test/java/org/folio/rest/impl/ProfileImportTest.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+      .when()
+      .get(JOB_PROFILES_PATH + "/" + jobProfileId)
+      .then()
+      .statusCode(HttpStatus.SC_OK);
+
+    async.complete();
+  }
+
+  @Test
+  public void shouldNotImportProfileSnapshotOfNotJobProfile(TestContext testContext) {
```
Comment (VRohach) [inline mod-di-converter-storage-server/src/test/java/org/folio/rest/impl/ProfileImportTest.java:195]: @RomanChernetskyi  TYPO: "shouldNotImportProfileSnapshotOfNotJobProfile" -> "shouldNotImportProfileSnapshotIfNotJobProfile".

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  @Override
+  public Future<ProfileSnapshotWrapper> importProfile(ProfileSnapshotWrapper profileSnapshot, String tenantId, OkapiConnectionParams okapiParams) {
+    if (profileSnapshot.getContentType() != JOB_PROFILE) {
+      String errorMessage = String.format(PROFILE_SNAPSHOT_INVALID_TYPE, profileSnapshot.getContentType(), JOB_PROFILE);
+      LOGGER.warn("importProfile:: {}", errorMessage);
+      return Future.failedFuture(new BadRequestException(errorMessage));
+    }
+    LOGGER.info("importProfile:: Started import for Job Profile with id {}", profileSnapshot.getProfileId());
```
Comment (Aliaksandr-Fedasiuk) [inline mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java:81]: Could you make this line the first line in this method?

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import static org.folio.rest.jaxrs.model.ProfileType.MAPPING_PROFILE;
+import static org.folio.rest.jaxrs.model.ProfileType.MATCH_PROFILE;
+import static org.folio.services.snapshot.ProfileSnapshotServiceImpl.convertProfileSnapshotWrapperContent;
+
+@Service
+public class ProfileImportServiceImpl implements ProfileImportService {
+  private static final Logger LOGGER = LogManager.getLogger();
+  private static final List<ProfileType> IMPORT_PROFILES_ORDER = List.of(MAPPING_PROFILE, ACTION_PROFILE, MATCH_PROFILE, JOB_PROFILE);
+  private static final String PROFILE_SNAPSHOT_INVALID_TYPE = "Cannot import profile snapshot of %s required type is %s";
+  private final EnumMap<ProfileType, BiFunction<ProfileSnapshotWrapper, OkapiConnectionParams, Future<Object>>> profileTypeToSaveFunction;
```
Comment (Aliaksandr-Fedasiuk) [inline mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java:47]: Perhaps the next names for profileTypeToSaveFunction variable will be clearer:
profileSaveFunctionsByProfileType
profileSaveHandlers
profileSaveStrategies?

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    if (snapshot.getChildSnapshotWrappers() != null) {
+      for (ProfileSnapshotWrapper child : snapshot.getChildSnapshotWrappers()) {
+        addSnapshotsToMap(child, profileTypeToSnapshots);
+      }
+    }
+  }
+
+  private List<ProfileAssociation> formAddedRelations(ProfileSnapshotWrapper rootSnapshot, ProfileType rootContentType) {
```
Comment (Aliaksandr-Fedasiuk) [inline mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java:131]: Please review this variant of the FormAddedRelations method.

```suggestion
private List<ProfileAssociation> formAddedRelations(ProfileSnapshotWrapper rootSnapshot, ProfileType rootContentType) {
    if (rootSnapshot.getChildSnapshotWrappers() == null) {
        return Collections.emptyList();
    }

    return rootSnapshot.getChildSnapshotWrappers().stream()
        .filter(childSnapshot -> shouldIncludeAssociation(rootContentType, childSnapshot))
        .flatMap(childSnapshot -> {
            List<ProfileAssociation> associations = new ArrayList<>();
            associations.add(createAssociation(rootSnapshot, childSnapshot));
            associations.addAll(formAddedRelations(childSnapshot, rootContentType));
            return associations.stream();
        })
        .collect(Collectors.toList());
}

private boolean shouldIncludeAssociation(ProfileType rootContentType, ProfileSnapshotWrapper childSnapshot) {
    return rootContentType == ACTION_PROFILE || childSnapshot.getContentType() != MAPPING_PROFILE;
}
```

**Review by Aliaksandr-Fedasiuk — CHANGES_REQUESTED** [review]
Please review the suggested improvements and add more log messages as well.
Thanks.

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java (hunk 1 above)
Comment (RomanChernetskyi) [reply mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java]: Done, please review

File: mod-di-converter-storage-server/src/test/java/org/folio/rest/impl/ProfileImportTest.java (hunk 2 above)
Comment (RomanChernetskyi) [reply mod-di-converter-storage-server/src/test/java/org/folio/rest/impl/ProfileImportTest.java]: Done, please review

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java (hunk 3 above)
Comment (RomanChernetskyi) [reply mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java]: Done, please review

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    } catch (Exception e) {
+      return Future.failedFuture(new BadRequestException(String.format("Cannot map profile content, error: %s", e.getMessage())));
+    }
+
+    var profileTypeToSnapshots = getProfileTypeToSnapshot(profileSnapshot);
+    var profileTypesInSaveOrder = profileTypeToSnapshots.keySet().stream()
+      .sorted(Comparator.comparingInt(IMPORT_PROFILES_ORDER::indexOf))
+      .toList();
+
+    Future<List<Object>> saveFuture = Future.succeededFuture();
```
[the remark this answers is not in the document]
Comment (RomanChernetskyi) [reply mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java]: Done, please review

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java (hunk 4 above)
Comment (RomanChernetskyi) [reply mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java]: Done, please review

**Review by Aliaksandr-Fedasiuk — APPROVED** [review]

File: mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java (hunk 5 above)
Comment (RomanChernetskyi) [reply mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java:140]: I think it is better to leave it as it is, thanks for your suggestion

**Review by VRohach — APPROVED** [review]

**Pull request merged**